_posts/2022-05-18-torchopt/torchopt.Rmd

```diff
@@ -0,0 +1,356 @@
+
+What a change! With ten-fold learning rate, the result is optimal. Does this mean the default setting is bad? Of course not; the algorithm has been tuned to work well with neural networks, not some function that has been purposefully designed to present a specific challenge.
+
+Naturally, we also have to see what happens for yet higher a learning rate.
+
+```{r}
+
+test_optim(
+    optim = optim_adamw,
+    # scale default rate by a factor of 70
+    opt_hparams = list(lr = 0.7),
+    test_fn = list(flower, function() (c(x0 = 20, y0 = 20, xmax = 3, xmin = -3, ymax = 3, ymin = -3))),
+    steps = 200
+)
+```
+
+![Minimizing the flower function with AdamW. Setup no. 3: lr = 0.7, 200 steps.](images/adamw-lr-0.7.png)
+
+We see the behavior we've always been warned about: Optimization hops around wildly, before seemingly heading off forever. (Seemingly, because in this case, this is not what happens. Instead, the search will jump far away, and back again, continuously.)
+
+Now, this might make one curious. What actually happens if we choose the "good" learning rate, but don't stop optimizing at two-hundred steps? Here, we try three-hundred instead:
+
+```{r}
+
+test_optim(
+    optim = optim_adamw,
+    # scale default rate by a factor of 10
+    opt_hparams = list(lr = 0.1),
+    test_fn = list(flower, function() (c(x0 = 20, y0 = 20, xmax = 3, xmin = -3, ymax = 3, ymin = -3))),
+    # this time, continue search until we reach step 300
+    steps = 300
+)
+```
+
+![Minimizing the flower function with AdamW. Setup no. 4: lr = 0.1, 300 steps.](images/adamw-lr-0.1-steps-300.png){alt="Minimizing the flower function with AdamW. Setup no. 3: lr "}
+
+Interestingly, we see the same kind of to-and-fro happening here as with a higher learning rate -- it's just delayed in time.
+
+Another playful question that comes to mind is: Can we track how the optimization process "explores" the four petals? With some quick experimentation, I arrived at this:
+
+![Minimizing the flower function with AdamW, lr = 0.1: Successive "exploration" of petals. Steps (clockwise): 300, 700, 900, 1300.](images/petals.png)
+
+Who says you need chaos[^4] to produce a beautiful plot?
+
+[^4]: Alluding to strange attractors here. If you're interested, you might want to check out the mini-series on chaos and deep learning: [Deep attractors: Where deep learning meets chaos](https://blogs.rstudio.com/ai/posts/2020-06-24-deep-attractors/), [Time series prediction with FNN-LSTM](https://blogs.rstudio.com/ai/posts/2020-07-20-fnn-lstm/), and [FNN-VAE for noisy time series forecasting](https://blogs.rstudio.com/ai/posts/2020-07-31-fnn-vae-for-noisy-timeseries/).
+
+## A second-order optimizer for neural networks: ADAHESSIAN
+
+On to the one algorithm I'd like to check out specifically. Subsequent to just a little bit of learning-rate experimentation, I was able to arrive at an excellent result after just thirty-five steps.
```

Comment:
```suggestion
On to the one algorithm I'd like to check out specifically. Subsequent to a little bit of learning-rate experimentation, I was able to arrive at an excellent result after just thirty-five steps.
```

Reply:
Since you have "just" in the following sentence, and already have a couple of qualifiers in the first, I cut it.

Reply:
yay :-)